radeontop.1
'\" t
.\"     Title: radeontop
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 03/13/2021
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "RADEONTOP" "1" "03/13/2021" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
radeontop \- tool to show GPU utilization
.SH "SYNOPSIS"
.sp
\fBradeontop [\-chmv] [\-b \fR\fB\fIbus\fR\fR\fB] [\-d \fR\fB\fIfile\fR\fR\fB] [\-l \fR\fB\fIlimit\fR\fR\fB] [\-i \fR\fB\fIsecs\fR\fR\fB] [\-p \fR\fB\fIdevice\fR\fR\fB] [\-t \fR\fB\fIticks\fR\fR\fB]\fR
.SH "DESCRIPTION"
.sp
RadeonTop shows the utilization of your GPU, both in general and by blocks\&.
.sp
Supported cards are R600 and up\&.
.SH "OPTIONS"
.PP
\fB\-b, \-\-bus \fR\fB\fIbus\fR\fR
.RS 4
If you have multiple cards, use this parameter to select a specific one\&. For lspci output
\fI01:00\&.0 VGA compatible controller\fR, bus would be
\fI1\fR\&.
.RE
.PP
\fB\-c, \-\-color\fR
.RS 4
Enables colorized output\&.
.RE
.PP
\fB\-d, \-\-dump \fR\fB\fIfile\fR\fR
.RS 4
Instead of displaying the data, dump it to
\fIfile\fR\&. Use
\fI\-\fR
for stdout\&. The file is appended to, so it\(cqs safe to specify the same filename for consecutive runs\&.

Use
\fISIGTERM\fR
or
\fISIGINT\fR
for graceful termination\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Displays help\&.
.RE
.PP
\fB\-i, \-\-dump\-interval \fR\fB\fIseconds\fR\fR
.RS 4
Wait for N seconds between data output\&.
.RE
.PP
\fB\-f, \-\-dump\-format \fR\fB\fIformat\fR\fR
.RS 4
Format to use for dumping, can be custom or json (default custom)\&. See DUMP FORMAT\&.
.RE
.PP
\fB\-l, \-\-limit \fR\fB\fIlimit\fR\fR
.RS 4
Quit after dumping
\fIlimit\fR
lines\&. Only valid in dump mode\&. Default is to dump until terminated; use
\fISIGTERM\fR
or
\fISIGINT\fR
for graceful termination\&.
.RE
.PP
\fB\-m, \-\-mem\fR
.RS 4
Forces the use of the /dev/mem path, in case autodetection fails\&. Useful for the proprietary driver\&.
.RE
.PP
\fB\-p, \-\-path \fR\fB\fIdevice\fR\fR
.RS 4
Open the DRM device node specified by
\fIdevice\fR
path\&. Usually it is located at /dev/dri/card0 or /dev/drm0\&. If this option used, \-\-bus and \-\-mem are ignored\&.
.RE
.PP
\fB\-t, \-\-ticks \fR\fB\fIticks\fR\fR
.RS 4
How many times per second to sample the data\&. The default is 120\&. Higher numbers mean more accurate output and more overhead\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Displays the program version\&.
.RE
.SH "INTERACTIVE COMMANDS"
.PP
\fBC, c\fR
.RS 4
Toggle colorized output\&.
.RE
.PP
\fBQ, q\fR
.RS 4
Quit\&.
.RE
.SH "DUMP FORMAT"
.sp
Fields tc, smx, cr, vram, gtt, mclk, sclk can be missing\&.
.sp
\fBcustom\fR
.sp
.if n \{\
.RS 4
.\}
.nf
<timestamp in sec: float>: bus <int>, gpu <Graphics pipe: percent>%, ee <Event Engine: percent>%, vgt <Vertex Grouper + Tesselator: percent>%, ta <Texture Addresser: percent>%, tc <Texture Cache: percent>%, sx <Shader Export: percent>%, sh <Sequencer Instruction Cache: percent>%, spi <Shader Interpolator: percent>%, smx <Shader Memory Exchange: percent>%, sc <Scan Converter: percent>%, pa <Primitive Assembly: percent>%, db <Depth Block: percent>%, cb <Color Block: percent>%, cr <Clip Rectangle: percent>%, vram <vram used: percent>% <vram used: float mb>mb <vram max: float mb>mb, gtt <gtt used: percent>% <gtt used: float mb>mb <gtt max: float mb>mb, mclk <Memory Clock: percent>% <Memory Clock: float ghz>ghz <Memory Clock max: float ghz>ghz, sclk <Shader Clock: percent>% <Shader Clock: float ghz>ghz <Shader Clock max: float ghz>ghz
.fi
.if n \{\
.RE
.\}
.sp
\fBjson\fR
.sp
.if n \{\
.RS 4
.\}
.nf
{"timestamp": <timestamp in sec: float>, "bus": <uint>, "gpu": <Graphics pipe: percent>, "ee": <Event Engine: percent>, "vgt": <Vertex Grouper + Tesselator: percent>, "ta": <Texture Addresser: percent>, "tc": <Texture Cache: percent>, "sx": <Shader Export: percent>, "sh": <Sequencer Instruction Cache: percent>, "spi": <Shader Interpolator: percent>, "smx": <Shader Memory Exchange: percent>, "sc": <Scan Converter: percent>, "pa": <Primitive Assembly: percent>, "db": <Depth Block: percent>, "cb": <Color Block: percent>, "cr": <Clip Rectangle: percent>, "vram": {"used_per": <vram used: percent>, "used_b": <vram used: uint bytes>, "max_b": <vram max: uint bytes>}, "gtt": {"used_per": <gtt used: percent>, "used_b": <gtt used: uint bytes>, "max_b": <gtt max: uint bytes>}, "mclk": {"used_per": <Memory Clock: percent>, "used_hz": <Memory Clock: float hz>, "max_hz": <Memory Clock max: float hz>}, "sclk": {"used_per": <Shader Clock: percent>, "used_hz": <Shader Clock: float hz>, "max_hz": <Shader Clock max: float hz>}}
.fi
.if n \{\
.RE
.\}
.SH "RESOURCES"
.sp
See http://github\&.com/clbr/radeontop


radeontop.c
/*
    Copyright (C) 2012 Lauri Kasanen
    Copyright (C) 2018 Genesis Cloud Ltd.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, version 3 of the License.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "radeontop.h"
#include <bits/getopt_core.h>
#include <getopt.h>

void die(const char * const why) {
	puts(why);
	exit(1);
}

static void version() {
	printf("RadeonTop %s\n", VERSION);
	exit(1);
}

static void help(const char * const me, const unsigned int ticks, const unsigned int dumpinterval, const enum DumpFormat dumpformat) {
	char * dumpformat_str = "";
	switch (dumpformat) {
		case Json:
			dumpformat_str = "json";
		break;
		case Custom:
			dumpformat_str = "custom";
		break;
	}

	printf(_("\n\tRadeonTop for R600 and above.\n\n"
		"\tUsage: %s [-chmv] [-b bus] [-d file] [-i seconds] [-l limit] [-p device] [-t ticks]\n\n"
		"-b --bus 3		Pick card from this PCI bus (hexadecimal)\n"
		"-c --color		Enable colors\n"
		"-d --dump file		Dump data to this file, - for stdout\n"
		"-i --dump-interval 1	Number of seconds between dumps (default %u)\n"
		"-f --dump-format custom	Foramt to use for dumping, can be `custom` or `json` (default %s)\n"
		"-l --limit 3		Quit after dumping N lines, default forever\n"
		"-m --mem		Force the /dev/mem path, for the proprietary driver\n"
		"-p --path device	Open DRM device node by path\n"
		"-t --ticks 50		Samples per second (default %u)\n"
		"-T --transparency	Enable transparency\n"
		"\n"
		"-h --help		Show this help\n"
		"-v --version		Show the version\n"),
		me, dumpinterval, dumpformat_str, ticks);
	die("");
}

int main(int argc, char **argv) {
	// Temporarily drop privileges to do option parsing, etc.
	seteuid(getuid());

	const unsigned int default_ticks = 120;
	const unsigned int default_dumpinterval = 1;
	const enum DumpFormat default_dumpformat = Custom;

	unsigned int ticks = default_ticks;
	unsigned char color = 0;
	unsigned char transparency = 0;
	short bus = -1;
	unsigned char forcemem = 0;
	unsigned int device_id = 0;
	unsigned int limit = 0;
	char *dump = NULL;
	unsigned int dumpinterval = default_dumpinterval;
	enum DumpFormat dumpformat = default_dumpformat;
	const char *path = NULL;

	// Translations
#ifdef ENABLE_NLS
	setlocale(LC_ALL, "");
	bindtextdomain("radeontop", "/usr/share/locale");
	textdomain("radeontop");
#endif

	// opts
	const struct option opts[] = {
		{"bus", 1, 0, 'b'},
		{"color", 0, 0, 'c'},
		{"dump", 1, 0, 'd'},
		{"dump-interval", 1, 0, 'i'},
		{"dump-format", 1, 0, 'f'},
		{"help", 0, 0, 'h'},
		{"limit", 1, 0, 'l'},
		{"mem", 0, 0, 'm'},
		{"path", 1, 0, 'p'},
		{"ticks", 1, 0, 't'},
		{"transparency", 0, 0, 'T'},
		{"version", 0, 0, 'v'},
		{0, 0, 0, 0}
	};

	while (1) {
		int c = getopt_long(argc, argv, "b:cTd:f:hi:l:mp:t:v", opts, NULL);
		if (c == -1) break;

		switch(c) {
			case 'h':
			case '?':
				help(argv[0], default_ticks, default_dumpinterval, default_dumpformat);
			break;
			case 't':
				ticks = atoi(optarg);
			break;
			case 'T':
				transparency = 1;
			break;
			case 'c':
				color = 1;
			break;
			case 'm':
				forcemem = 1;
			break;
			case 'b':
				bus = strtoul(optarg, NULL, 16);
			break;
			case 'v':
				version();
			break;
			case 'l':
				limit = atoi(optarg);
			break;
			case 'd':
				dump = optarg;
			break;
			case 'i':
				dumpinterval = atoi(optarg);
				if (dumpinterval < 1)
					dumpinterval = 1;
			break;
			case 'f':
				if (strcmp(optarg, "custom") == 0) {
					dumpformat = Custom;
				} else if (strcmp(optarg, "json") == 0) {
					dumpformat = Json;
				} else {
					die("Unknown dump format");
				}
			break;
			case 'p':
				path = optarg;
			break;
		}
	}

	// init (regain privileges for bus initialization and ultimately drop them afterwards)
	seteuid(0);
	init_pci(path, &bus, &device_id, forcemem);
	// after init_pci we can assume that bus/device_id exists (otherwise it would die())

	setuid(getuid());

	const int family = getfamily(device_id);
	if (!family)
		puts(_("Unknown Radeon card. <= R500 won't work, new cards might."));

	const char * const cardname = family_str[family];

	initbits(family);

	// runtime
	collect(ticks, dumpinterval);

	if (dump)
		dumpdata(ticks, dump, limit, bus, dumpinterval, dumpformat);
	else
		present(ticks, cardname, color,transparency, bus, dumpinterval);

	cleanup();
	return 0;
}
